### PR TITLE
Bind provider form to api call

### PR DESCRIPTION
**Pull request by fredx30**
This is a mirror implementation of the code that binds gitlab/azuredevops to the wahrf-web interface. This uses the wharf-github-provider to import the whaf config. 

### To use-
To add the wharf-api project the Github provider from can be use in the following way.
Post to `/import/github`
```json
{
    "group": "iver-wharf",
    "project": "wharf-api",
    "token": "${GITHUB-TOKEN}",
    "url": "https://api.github.com"
} 
```
image

Links to other relevant PRs:
- https://github.com/iver-wharf/wharf-provider-github/pull/10

### CODE REVIEW

File: src/app/projects/project-list/project-list-item.component.ts [hunk 1]
```diff
@@ -29,9 +29,13 @@ export class ProjectListItemComponent implements OnChanges {
...
-      this.providerHostname = this.project.provider?.url
-        ? new URL(this.project.provider.url).hostname
-        : null;
+      try {
+        this.providerHostname = this.project.provider?.url
+          ? new URL(this.project.provider.url).hostname
+          : null;
+      } catch (e) {
+        this.providerHostname = null;
+      }
```
Comment (jilleJr) [inline src/app/projects/project-list/project-list-item.component.ts:38]: This seems to come from #6 . Suggest marking that PR as "Ready for Review" or add "Closes #6" on this PR

File: src/app/providers/provider-form/github/github.component.ts [hunk 2]
```diff
@@ -11,13 +14,24 @@ export class GithubComponent {
...
+      token: this.providerForm.value.token,
+      group: this.providerForm.value.group,
+      project: this.providerForm.value.project,
+      uploadUrl: this.providerForm.value.uploadUrl
+    };
+    this.gitHubService.githubPost(providerData)
+      .pipe(
+        finalize(() => this.providersService.triggerCloseForm(this.providerForm))
+      )
+      .subscribe();
```
Comment (jilleJr) [inline src/app/providers/provider-form/github/github.component.ts:35]: nit: Does this need to unsubscribe when the component is destroyed? Such as if the user clicks close and then opens the model again before the request has finished

File: src/app/providers/providers.service.ts [hunk 3]
```diff
@@ -37,6 +43,16 @@ export class ProvidersService {
...
+    } else if (project.provider.name === ProviderType.GitHub.toLowerCase()) {
+      const importData: GitHubMainImport = {
+        url: project.provider.url,
+        tokenId: project.provider.tokenId,
+        group: project.groupName,
+        project: project.name,
+        providerId: project.provider.providerId,
+        uploadUrl: project.provider.uploadUrl,
+      };
+      return this.gitHubService.githubPost(importData);
```
Comment (jilleJr) [inline src/app/providers/providers.service.ts:55]: nit: Suggest changing this to a `switch` statement

**Review by jilleJr — APPROVED** [review]
Some nitpick suggestions, meaning it's OK to leave it as-is

**Review by iverestefans — APPROVED** [review]

**Review by jilleJr — APPROVED** [review]

File: src/app/projects/project-list/project-list-item.component.ts (hunk 1 above)
Comment (fredx30) [reply src/app/projects/project-list/project-list-item.component.ts]: I have a cool idea for resolution on this- i think input can neatly be sanitised for this purpose with a custom validator. Will continue this change in #6 and will therefore drop it on this PR.

File: src/app/providers/provider-form/github/github.component.ts (hunk 2 above)
Comment (fredx30) [reply src/app/providers/provider-form/github/github.component.ts]: I rewrote this part to a `pipe(first())` which will automagically handle the unsub on the response reception. Also after further consideration i moved the close logic for the providerForm to the subscribe. That way the window will not close on a sub error. See https://github.com/iver-wharf/wharf-web/pull/8/commits/5e89ca3940fd4b8f65ca52482008552d648c4963

File: src/app/providers/providers.service.ts (hunk 3 above)
Comment (fredx30) [reply src/app/providers/providers.service.ts:55]: I wrote this version 
```ts
  refreshProject(project: WharfProject): Observable<any> {
    let importData: GitHubMainImport | GitlabMainImport | AzureDevOpsMainImport;
    switch (project.provider.name) {
      case ProviderType.GitLab.toLowerCase():
        importData = {
          url: project.provider.url,
          tokenId: project.provider.tokenId,
          group: project.groupName,
          project: project.name,
          providerId: project.provider.providerId
        };
        return this.gitlabService.gitlabPost(importData);
      case ProviderType.GitHub.toLowerCase():
        importData = {
          url: project.provider.url,
          tokenId: project.provider.tokenId,
          group: project.groupName,
          project: project.name,
          providerId: project.provider.providerId,
          uploadUrl: project.provider.uploadUrl,
        };
        return this.gitHubService.githubPost(importData);
      case ProviderType.AzureDevOps.toLowerCase():
        importData = {
          url: project.provider.url,
          tokenId: project.provider.tokenId,
          group: project.groupName,
          project: project.name,
          providerId: project.provider.providerId
        };
        return this.azureDevOpsService.azuredevopsPost(importData);
    }
  }
```

Because of the nature of the block the const cant be redeclared though so we would loose the type saftey it seems. I vote it stays as is.

File: src/app/providers/provider-form/github/github.component.ts [hunk 4]
```diff
@@ -11,13 +14,24 @@ export class GithubComponent {
...
+      url: this.providerForm.value.url,
+      token: this.providerForm.value.token,
+      group: this.providerForm.value.group,
+      project: this.providerForm.value.project,
+    };
+    this.gitHubService.githubPost(providerData)
+      .pipe(first())
+      .subscribe(
+        success => this.providersService.triggerCloseForm(this.providerForm),
+        error => {/* Error gets show to screen form window does not close. */},
```
Comment (Pikabanga) [inline src/app/providers/provider-form/github/github.component.ts:34]: I don't understand what this comment means. Can you rephrase it?

Anyway, I don't think you need this empty func, you can just remove this error callback if it's ignored anyway.

**Review by Pikabanga — APPROVED** [review]

**Review by jilleJr — CHANGES_REQUESTED** [review]
I almost forgot! Please make sure to update the `CHANGELOG.md` file

File: src/app/providers/providers.service.ts (hunk 3 above)
Comment (fredx30) [reply src/app/providers/providers.service.ts]: Taken! Thanks.

File: src/app/providers/provider-form/github/github.component.ts (hunk 4 above)
Comment (fredx30) [reply src/app/providers/provider-form/github/github.component.ts]: I removed it. 
The comment should say something along the lines of- when the post for the github provider completes successfully the form is automatically cleared and closed. When the post to the github provider does not complete sucessfully then the form is not closed and all the field values are retained.

**Comment by fredx30** [comment]
> I almost forgot! Please make sure to update the `CHANGELOG.md` file

Changelog.md is now also updated.

File: CHANGELOG.md [hunk 5]
```diff
@@ -12,6 +12,11 @@ This project tries to follow [SemVer 2.0.0](https://semver.org/).
 	https://changelog.md/
 -->
 
+## v1.2.1 (WIP)
+
+- Add logic to the empty GitHub post service. This will now post form contents to the github-provider.
+- Add refresh logic for the GitHub configurations.
```
Comment (jilleJr) [inline CHANGELOG.md:18]: We've a bunch of conventions here that I haven't explained.

- One is that the changelogs shall start with one of the verbs from https://changelog.md, so suggest changing "Add" &rarr; "Added".

- Then since this is added features, SemVer2.0.0 states that this should be a minor version bump. Patch version bumps are for bug fixes. The "(WIP)" badge is spot on though!

- A PR reference should be added to the end. You can see lots of them have `!94` syntax, which is for GitLab. For GitHub we do `#8`.

- We apply `remark-lint` checks here, and one of the checks as Codacy states here is that it must be wrapped if it spans longer than 80 characters. You can run the linting locally by doing `npm i` and `npm run lint`

So all in all:

```suggestion
## v1.3.0 (WIP)

- Added logic to the empty GitHub post service. This will now post form
  contents to the github-provider. (#8)

- Added refresh logic for the GitHub configurations. (#8)
```

File: src/app/providers/providers.service.ts [hunk 6]
```diff
@@ -21,31 +22,41 @@ export class ProvidersService {
 
   constructor(
     private gitlabService: GitlabService,
-    private azureDevOpsService: AzureDevOpsService) { }
+    private gitHubService: GitHubService,
+    private azureDevOpsService: AzureDevOpsService) {
+  }
```
Comment (jilleJr) [inline src/app/providers/providers.service.ts:27]: nit: In TS I personally like to add comma at the end of even the last argument just so the Git diffs are more focused on what actually changed.

Also, then you can sort the parameter lines

```suggestion
    private azureDevOpsService: AzureDevOpsService,
  ) { }
```

**Review by jilleJr — CHANGES_REQUESTED** [review]
Just some changes in CHANGELOG.md. Otherwise everything is spot on!

File: src/app/providers/providers.service.ts (hunk 6 above)
Comment (fredx30) [reply src/app/providers/providers.service.ts:27]: Ill make a separate PR to deal with this by using eslints `"comma-dangle": ["error", "always-multiline"]` as well.

**Review by jilleJr — APPROVED** [review]
LGTM!
